controllers/nstemplateset/space_roles_test.go

```diff
@@ -0,0 +1,210 @@
+				withTemplateRefUsingRevision("abcde10"), // starting with an older revision
+				withWorkspaceLabel("oddity"))
+			mgr, memberClient := prepareSpaceRolesManager(t, nsTmplSet, ns)
+
+			// when
+			createdOrUpdated, err := mgr.ensure(logger, nsTmplSet)
+
+			// then
+			require.NoError(t, err)
+			assert.True(t, createdOrUpdated)
+			// verify that roles and rolebindings were created
+			AssertThatRole(t, "oddity-appstudio", "space-admin", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			AssertThatRoleBinding(t, "oddity-appstudio", "user1-space-admin", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			AssertThatRoleBinding(t, "oddity-appstudio", "user2-space-admin", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			AssertThatRole(t, "oddity-appstudio", "space-viewer", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			AssertThatRoleBinding(t, "oddity-appstudio", "user3-space-viewer", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			AssertThatRoleBinding(t, "oddity-appstudio", "user4-space-viewer", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			// verify that the `last-applied-space-roles` annotation was set on namespace
+			lastApplied, err := json.Marshal(nsTmplSet.Spec.SpaceRoles)
+			require.NoError(t, err)
+			AssertThatNamespace(t, "oddity-appstudio", memberClient.Client).
+				HasAnnotation(toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey, string(lastApplied))
+		})
+
+		t.Run("update roles", func(t *testing.T) {
+
+			t.Run("add admin user", func(t *testing.T) {
+				// given
+				nsTmplSet := newNSTmplSet(commontest.MemberOperatorNs, "oddity", "appstudio",
+					withSpaceRoles(map[string][]string{
+						"admin-appstudio-abcde11": {"user1", "user2"},
+					}))
+				ns := newNamespace(nsTmplSet.Spec.TierName, "oddity", "appstudio", // ns.name=oddity=appstudio
+					withTemplateRefUsingRevision("abcde10"), // starting with an older revision
+					withWorkspaceLabel("oddity"),
+				)
+				mgr, memberClient := prepareSpaceRolesManager(t, nsTmplSet, ns)
+				_, err := mgr.ensure(logger, nsTmplSet)
```

Comment:
could you please add a comment mentioning why the ensure is called here? and could you please add it to the following ones as well?
```suggestion
				// running ensure to precreate the resources for the initial set of SpaceRoles
				_, err := mgr.ensure(logger, nsTmplSet)
```

Reply:
done in a1919ee